src/Mvc/Mvc.Razor/src/RazorHotReload.cs

```diff
@@ -0,0 +1,68 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.AspNetCore.Mvc.ApplicationParts;
+using Microsoft.AspNetCore.Mvc.Razor.Compilation;
+using Microsoft.AspNetCore.Mvc.Razor.TagHelpers;
+
+namespace Microsoft.AspNetCore.Mvc.Razor
+{
+    internal sealed class RazorHotReload
+    {
+        private readonly RazorCompiledItemFeatureProvider? _razorCompiledItemFeatureProvider;
+        private readonly DefaultViewCompiler? _defaultViewCompiler;
+        private readonly RazorViewEngine? _razorViewEngine;
+        private readonly RazorPageActivator? _razorPageActivator;
+        private readonly DefaultTagHelperFactory? _defaultTagHelperFactory;
+        private readonly TagHelperComponentPropertyActivator? _tagHelperComponentPropertyActivator;
+
+        public RazorHotReload(
+            IRazorViewEngine razorViewEngine,
+            IRazorPageActivator razorPageActivator,
+            ITagHelperFactory tagHelperFactory,
+            IViewCompilerProvider viewCompilerProvider,
+            ITagHelperComponentPropertyActivator tagHelperComponentPropertyActivator,
+            ApplicationPartManager applicationPartManager)
```

Comment:
Do we need to do anything for page action descriptors? (I imagine at some point the endpoints need to be recalculated)

Reply:
The `CompiledPageActionDescriptorProvider` uses the ApplicationPart feature to get the set of Razor Pages - https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.RazorPages/src/Infrastructure/CompiledPageActionDescriptorProvider.cs#L56-L61. It's a fairly similar code path to how the RazorViewEngine calculates it's inputs. We update that as part of this type, and poke the IActionDescriptorChangeProvider as part of `HotReloadService`, so it causes Razor Pages and it's routing to be rediscovered.

It might be worthwhile to create a fake provider and simulate this e2e given how many moving parts there are to this. I'd be happy to do that in a follow up (just getting the runtime changes in since it's a major hassle to sync across dotnet/aspnetcore, dotnet/sdk, and dotnet/roslyn to test these changes out)